### PR TITLE
Support saturating instructions

### PR DESCRIPTION
**Pull request by pranavk**
This is a draft pull request. Work still in progress. I need to add more tests. Right now only one small test to test the basic functionality (which works).

### CODE REVIEW

**Comment by regehr** [comment]
great!
I don't want to merge until llvm 8 is really out
if you have time in the meantime, add the other saturating instructions
I'll look over your code now

File: lib/Extractor/Candidates.cpp [hunk 1]
```diff
@@ -563,6 +563,11 @@ Inst *ExprBuilder::buildHelper(Value *V) {
           Inst *Overflow = IC.getInst(Inst::UMulO, 1, {L, R}, /*Available=*/false);
           return IC.getInst(Inst::UMulWithOverflow, L->Width+1, {Mul, Overflow});
         }
+        case Intrinsic::sadd_sat: {
+	  // TODO: Factor out *R
+	  Inst *R = get(II->getOperand(1));
+	  return IC.getInst(Inst::SAddSat, L->Width, {L, R}, /*Available=*/false);
```
Comment (regehr) [inline lib/Extractor/Candidates.cpp]: why false for available?

**Comment by regehr** [comment]
ok code looks great except:
- thing I noted
- we require parser support
- need quite a few more tests, think like Roman

**Comment by regehr** [comment]
ok I see you made a change to the parser -- is this all that's required?

File: lib/Extractor/Candidates.cpp (hunk 1 above)
Comment (pranavk) [reply lib/Extractor/Candidates.cpp]: i am not sure really yet how it affects things. Yeah, from the pattern of other instructions above, it should be true (which is also the default if i don't mention it).

**Comment by pranavk** [comment]
the one test that i added passed; so, i guess this is all that's required in the parser to make it work. i'll fix things as i notice them while adding newer tests

**Comment by regehr** [comment]
look at @LebedevRI's tests for the funnel shift patch
your tests do not need to be as extensive as his (where were excellent) but you should cover that same sorts of things

**Comment by pranavk** [comment]
added rest of the instructions; will add tests soon too

File: lib/Extractor/Candidates.cpp (hunk 1 above)
Comment (regehr) [reply lib/Extractor/Candidates.cpp]: should be true

Comment (pranavk) [reply lib/Extractor/Candidates.cpp]: this is outdated. i have already fixed that.

**Comment by regehr** [comment]
ready for review?

File: lib/Extractor/KLEEBuilder.cpp [hunk 2]
```diff
@@ -366,6 +366,50 @@ class KLEEBuilder : public ExprBuilder {
...
+      auto smaxExt = SExtExpr::create(smax, I->Width + 1);
+      auto pred = SleExpr::create(addExt, sminExt);
+
+      auto pred2 = SgeExpr::create(addExt, smaxExt);
+      auto select2 = SelectExpr::create(pred2, smax, add);
+
+      return SelectExpr::create(pred, smin, select2);
+    }
+    case Inst::UAddSat:
+    {
```
Comment (regehr) [inline lib/Extractor/KLEEBuilder.cpp]: prefer putting this kind of curly brace on the same line as the case
(don't worry about it for this PR but do it generally)

**Comment by regehr** [comment]
ok this seems almost ready to go

I want at least one test case which synthesizes a saturating operation

I want at least one test case which reasons about saturating arithmetic symbolically, for example use a path condition to express that an addition cannot overflow, in which case you should show that a saturating subtract can be replaced by a subnsw

**Comment by pranavk** [comment]
ok. will add these tests and update you when it's ready

File: lib/Extractor/KLEEBuilder.cpp [hunk 3]
```diff
@@ -367,6 +367,46 @@ class KLEEBuilder : public ExprBuilder {
...
+      ref<Expr> add = AddExpr::create(get(Ops[0]), get(Ops[1]));
+      auto sextL = SExtExpr::create(get(Ops[0]), I->Width + 1);
+      auto sextR = SExtExpr::create(get(Ops[1]), I->Width + 1);
+      auto addExt = AddExpr::create(sextL, sextR);
+      auto smin = klee::ConstantExpr::alloc(llvm::APInt::getSignedMinValue(I->Width));
+      auto smax = klee::ConstantExpr::alloc(llvm::APInt::getSignedMaxValue(I->Width));
+      auto sminExt = SExtExpr::create(smin, I->Width + 1);
+      auto smaxExt = SExtExpr::create(smax, I->Width + 1);
+      auto pred = SleExpr::create(addExt, sminExt);
+
```
Comment (regehr) [inline lib/Extractor/KLEEBuilder.cpp]: I feel like this is a pretty complicated way to detect signed overflow. why not reuse our signed overflow code or else just look for the result having a different sign than the inputs (you don't need to do any sign extension in that case)

File: test/Pass/syn-inst-sadd-sat.ll [hunk 4]
```diff
@@ -0,0 +1,19 @@
+; REQUIRES: solver
```
Comment (regehr) [inline test/Pass/syn-inst-sadd-sat.ll]: you should try to add diversity in test cases
if the souper test used one way to detect overflow, use a different one here
if the souper test is 8 bits, make this one 16 or 32
etc.
you're trying to dig up bugs here, doing different things in different ways is part of the goal

File: test/Tool/saturating1.opt [hunk 5]
```diff
@@ -0,0 +1,10 @@
+; REQUIRES: solver
+
+; RUN: %souper-check %solver -infer-rhs -print-replacement %s > %t1
+; RUN: %FileCheck %s -check-prefix=SUCCESS < %t1
+
+; SUCCESS: RHS inferred successfully
```
Comment (regehr) [inline test/Tool/saturating1.opt]: do you want to look for a particular RHS? or are you just testing that souper thinks it has succeeded here?

File: lib/Extractor/KLEEBuilder.cpp (hunk 3 above)
Comment (pranavk) [reply lib/Extractor/KLEEBuilder.cpp]: this (and others) is taken verbatim from alive2 to keep it consistent with it; I didn't want to come up with any clever way for this (with potential risks of getting wrong). Moreover, we want to get rid of KLEE anyways according to you; so, i didn't want to invest much time on it.

File: test/Tool/saturating1.opt (hunk 5 above)
Comment (pranavk) [reply test/Tool/saturating1.opt]: yeah, this test is silly. I will remove it. I think we have enough tests for this one now.

Comment (pranavk) [reply test/Tool/saturating1.opt]: fixed

File: lib/Extractor/KLEEBuilder.cpp (hunk 3 above)
Comment (regehr) [reply lib/Extractor/KLEEBuilder.cpp]: ya, ok ok :)

**Comment by regehr** [comment]
LGTM, thanks!

**Comment by regehr** [comment]
do you know what's going on with the test failures?

**Comment by pranavk** [comment]
i have seen this KLEE buidler recursively calling itself repeateadly in these tests before. So this should be unrelated to this PR.

**Comment by pranavk** [comment]
ok. these tests fails on my local even with this PR reverted. So this is unrelated.

**Comment by LebedevRI** [comment]
(That's just #400)

**Comment by regehr** [comment]
this is a problem that'll go away once we move to alive!
not sure what to do about it right now, ugh. I'll merge this though.

**Comment by regehr** [comment]
@LebedevRI but I don't see why it's showing up in CI suddenly, ugh....

**Pull request merged**